Mathlib/Analysis/Calculus/ImplicitContDiff.lean

```diff
@@ -0,0 +1,142 @@
+/-
+Copyright (c) 2025 Winston Yin. All rights reserved.
+Released under Apache 2.0 license as described in the file LICENSE.
+Authors: Winston Yin
+-/
+import Mathlib.Analysis.Calculus.Implicit
+import Mathlib.Analysis.Calculus.InverseFunctionTheorem.ContDiff
+
+/-!
+# Implicit function theorem
+
+In this file, we apply the generalised implicit function theorem to the more familiar case and show
+that the implicit function preserves the smoothness class of the implicit equation.
+
+Let `E` and `F` be real or complex Banach spaces. Let `f : E × F → F` be a function that is $C^n$ at
+a point `(a, b) : E × F`, where `n ≥ 1`. Let `f'` be the derivative of `f` at `(a, b)`. If the range
+of `f'` is all of `F`, and the kernel of `f'` is the subspace `E × {0}` in `E × F`, then there
+exists a function `φ : E → F` such that `φ a = b`, and `f x (φ x) = f a b` holds for all `x` in a
+neighbourhood of `a`. Furthoremore, `φ` is $C^n$ at `a`.
+
+## TODO
+* Local uniqueness of the implicit function
+* Derivative of the implicit function
+
+## Tags
+
+implicit function, inverse function
+-/
+
+namespace ImplicitFunctionData
+
+variable {𝕜 : Type*} [RCLike 𝕜] {E : Type*} [NormedAddCommGroup E]
+  [NormedSpace 𝕜 E] [CompleteSpace E] {F : Type*} [NormedAddCommGroup F] [NormedSpace 𝕜 F]
+  [CompleteSpace F] {G : Type*} [NormedAddCommGroup G] [NormedSpace 𝕜 G] [CompleteSpace G]
+  (φ : ImplicitFunctionData 𝕜 E F G) {n : WithTop ℕ∞}
+
+/-- The implicit function defined by a $C^n$ implicit equation is $C^n$. This applies to the general
+form of the implicit function theorem. -/
+theorem contDiff_implicitFunction (hl : ContDiffAt 𝕜 n φ.leftFun φ.pt)
+    (hr : ContDiffAt 𝕜 n φ.rightFun φ.pt) (hn : 1 ≤ n) :
+    ContDiffAt 𝕜 n φ.implicitFunction.uncurry (φ.prodFun φ.pt) := by
+  rw [implicitFunction, Function.uncurry_curry, toOpenPartialHomeomorph,
+    ← HasStrictFDerivAt.localInverse_def]
+  exact (hl.prodMk hr).to_localInverse (φ.hasStrictFDerivAt |>.hasFDerivAt) hn
+
+end ImplicitFunctionData
+
+open Filter
+
+open LinearMap (ker range)
+
+open scoped Topology
+
+/-- A predicate stating the sufficient conditions on an implicit equation `f : E × F → F` that will
+lead to a $C^n$ implicit function `φ : E → F`. -/
+structure IsContDiffImplicitAt {𝕜 : Type*} [RCLike 𝕜]
```

Comment:
This repeats the variables below. Mathlib is not always consistent about this; I tend to at least put k, E and F outside. (Keeping n, f, f' and a sounds good --- and hints at the other being useful.) What do you think?

In general, I agree that there is a lot of value in including all necessary hypotheses in a definition, but the type of f' implies these (except for completeness, see below).